components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodPortsService.kt

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote
+
+import com.intellij.openapi.diagnostic.thisLogger
+import com.jetbrains.rd.util.URI
+import org.apache.http.client.utils.URIBuilder
+import java.util.Optional
+import java.util.regex.Pattern
+
+class GitpodPortsService {
+    companion object {
+        /** Host used by forwarded ports on JetBrains Client. */
+        const val FORWARDED_PORT_HOST = "127.0.0.1"
+    }
+    private val hostToClientForwardedPortMap: MutableMap<Int, Int> = mutableMapOf()
+
+    fun isForwarded(hostPort: Int): Boolean = hostToClientForwardedPortMap.containsKey(hostPort)
+
+    private fun getForwardedPort(hostPort: Int): Optional<Int> = Optional.ofNullable(hostToClientForwardedPortMap[hostPort])
+
+    fun setForwardedPort(hostPort: Int, clientPort: Int) {
+        hostToClientForwardedPortMap[hostPort] = clientPort
+    }
+
+    fun removeForwardedPort(hostPort: Int) {
+        hostToClientForwardedPortMap.remove(hostPort)
+    }
+
+    fun getLocalHostUriFromHostPort(hostPort: Int): URI {
+        val optionalForwardedPort = getForwardedPort(hostPort)
+
+        val port = if (optionalForwardedPort.isPresent) {
+            optionalForwardedPort.get()
+        } else {
+            thisLogger().warn(
+                    "gitpod: Tried to get the forwarded port of $hostPort, which was not forwarded. " +
+                    "Returning $hostPort itself."
+            )
+            hostPort
+        }
+
+        return URIBuilder()
+                .setScheme("http")
+                .setHost(FORWARDED_PORT_HOST)
```

Comment:
If the port was not forwarded, will it worked in `127.0.0.1`? If not (if they are works with workspace URL), the `scheme` should change to `https` for gitpod SaaS too.

(self-hosted with self build may make it possible to use http themselves)

Reply:
Gotcha!
I confirm `getLocalHostUriFromHostPort` function is only used for getting the local URI after checking it's being forwarded (by using `isForwarded` function).
Outside this class, we're handling the non-forwarded cases and properly using `https`.